### PR TITLE
Added dispatching of the checkout dedicated events

### PR DESCRIPTION
**Pull request by Tetragramat**
Checkout dedicated events doc: https://github.com/Sylius/Sylius/blob/master/docs/book/orders/checkout.rst#checkout-related-events

fixes issue #250

### CODE REVIEW

**Comment by Tetragramat** [comment]
@lchrusciel could someone look at this PR?

**Comment by Tetragramat** [comment]
Thank you for your answer @lchrusciel .
I've looked into those events again.
Events:
`sylius.order.pre_address`
`sylius.order.post_address`
`sylius.order.pre_select_shipping`
`sylius.order.post_select_shipping`
`sylius.order.pre_payment`
`sylius.order.post_payment`
Doesn't do currently anything except redirect in post events.
Event `sylius.order.pre_complete` sets ip and locale into order and checks again consistency of the order. Locale is already set into order by plugin, but ip is not.
Event `sylius.order.post_complete` sends order confirmation email, which is not send by plugin.

So If you do not like triggering the events, then do you think adding submitting the confirmation email would be ok? I think that client ip should be set too.

There is still posibility that when someone makes plugin with listeners listening on checkout events, then that plugin will not work correctly with this plugin because of missing event triggers, but at the same time that plugin may break shop api. So I understand.

**Comment by lchrusciel** [comment]
You are right about last two. Assigning an IP address should be a part of order completion I suppose. With the second one, the answer requires much more work because it seems to be a much better solution to send an event over event bus, so anyone can listen to it(and one of the listeners would send an email). Of course, it should be added to all of the handlers, but I do not have time for it right now :( I could be as simple as dispatching an event inside of handler as the last action, no need of introducing ES. Do you think you can add it to the plugin? 

If no, as a shortcut we can just explicitly call an email service from the controller which dispatch complete command.

**Comment by Tetragramat** [comment]
@lchrusciel I got time now, so I can do dispatching of that event. There is a litle problem. I'm not sure I understand you corrently. What you're asking for is dispatching of an event as the last action in every handler and after that make listener on one of those events which will send an email?
I do not want to redo it multiple times, because I might have no time again like in the last two weeks.

File: src/Handler/AddressOrderHandler.php [hunk 1]
```diff
@@ -81,6 +91,10 @@ public function handle(AddressOrder $addressOrder)
         $order->setShippingAddress($shippingAddress);
         $order->setBillingAddress($billingAddress);
 
+        $this->eventDispatcher->dispatch('sylius.order.pre_address', new ResourceControllerEvent($order));
```
Comment (lchrusciel) [inline src/Handler/AddressOrderHandler.php]: Missing specs for dispatching commands

Comment (Tetragramat) [reply src/Handler/AddressOrderHandler.php]: Sorry, this is second time I saw phpspec and I'm unable to figure to how to spec this line.

**Comment by mamazu** [comment]
Is there an update to this?

**Comment by Tetragramat** [comment]
@mamazu I do not plan to continue working on this PR. Anyone can take it and finish it.

**Comment by mamazu** [comment]
Okay, I created #312 but thanks for the foundations.

**Comment by Zales0123** [comment]
Can be closed due to #312

**Pull request closed without merging**